plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergPageSourceProvider.java

```diff
@@ -356,6 +378,102 @@ private static ConnectorPageSource createOrcPageSource(
+     */
+    public static Optional<ReaderColumns> projectColumns(List<IcebergColumnHandle> columns)
+    {
+        // No projection is required if all columns are base columns
+        if (columns.stream().allMatch(IcebergColumnHandle::isBaseColumn)) {
+            return Optional.empty();
+        }
+
+        Map<IcebergColumnHandle, Integer> originalColumnIndexes = Streams.mapWithIndex(columns.stream(), Map::entry)
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> toIntExact(entry.getValue())));
+        // Order the columns in path depth order so existing projections for nested columns can be reused
+        List<IcebergColumnHandle> orderedColumns = originalColumnIndexes.keySet().stream()
+                .sorted(Comparator.comparingInt(col -> col.getPath().size()))
+                .collect(toImmutableList());
+
+        ImmutableList.Builder<ColumnHandle> projectedColumns = ImmutableList.builder();
+        Integer[] outputColumnMapping = new Integer[columns.size()];
+        Map<List<String>, Integer> mappedColumnIndices = new HashMap<>();
+        int projectedColumnIndex = 0;
+
+        for (IcebergColumnHandle column : orderedColumns) {
+            Optional<Integer> existingProjectedColumn = findExistingProjection(mappedColumnIndices, column);
+            int columnIndex = originalColumnIndexes.get(column);
+            if (existingProjectedColumn.isPresent()) {
+                outputColumnMapping[columnIndex] = existingProjectedColumn.get();
+                continue;
+            }
+
+            projectedColumns.add(column);
+            mappedColumnIndices.put(column.getFullPath(), projectedColumnIndex);
+            outputColumnMapping[columnIndex] = projectedColumnIndex;
+            projectedColumnIndex++;
+        }
+
+        return Optional.of(new ReaderColumns(projectedColumns.build(), ImmutableList.copyOf(outputColumnMapping)));
+    }
+
+    private static Optional<Integer> findExistingProjection(Map<List<String>, Integer> mappedColumnIndices, IcebergColumnHandle column)
+    {
+        // TODO: This is n^2 on the number of columns
+        for (Map.Entry<List<String>, Integer> existingMapping : mappedColumnIndices.entrySet()) {
+            List<String> existingPath = existingMapping.getKey();
+            if (column.getPath().size() >= existingPath.size() && column.getPath().subList(0, existingPath.size()).equals(existingPath)) {
+                // A parent of the column is already projected, so we don't have to read it again
+                return Optional.of(existingMapping.getValue());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private static Map<Integer, OrcColumn> mapIdsToOrcFileColumns(List<OrcColumn> columns)
+    {
+        ImmutableMap.Builder<Integer, OrcColumn> columnMap = ImmutableMap.builder();
+        columns.stream()
+                .filter(orcColumn -> orcColumn.getAttributes().containsKey(ORC_ICEBERG_ID_KEY))
```

Comment:
Is it required that `ORC_ICEBERG_ID_KEY` will be present for each column? What happens if it is not?

Reply:
Thanks for the reminder. I think that case will likely either fail or return incorrect results. I was trying to find an existing test that covers that case, but I haven't been able to figure out when that property will be missing.

Reply:
I was looking through the Iceberg spec document and it seems to suggest that this should never happen. 

> Columns in Iceberg data files are selected by field id. The table schema’s column names and order may change after a data file is written, and projection must be done using field ids. If a field id is missing from a data file, its value for each row should be null.

https://iceberg.apache.org/spec/#column-projection